Add modulemd-stream v3

mmathesius opened the pull request:
Initial implementation of new ModuleStreamV3 object, parsing/emitting of `modulemd-stream` v3 documents, and a basic parse/dump test.

mmathesius commented:
CC: @mcurlej

sgallagher commented on modulemd/modulemd-module-stream.c:
```diff
@@ -472,7 +494,8 @@ modulemd_module_stream_upgrade (ModulemdModuleStream *self,
   if (!mdversion)
     {
       /* If target mdversion is unspecified, set it to the latest */
-      mdversion = MD_MODULESTREAM_VERSION_LATEST;
+      /* TODO: reset to MD_MODULESTREAM_VERSION_LATEST */
+      mdversion = MD_MODULESTREAM_VERSION_TWO;
     }
```
Should this TODO be updated?

mmathesius replied:
Yes, thanks for catching the temporary tweak that can now be reverted.

sgallagher commented on modulemd/modulemd-module-stream.c:
```diff
@@ -1179,7 +1202,8 @@ modulemd_module_stream_class_init (ModulemdModuleStreamClass *klass)
     "Metadata Version",
     "The metadata version of this ModuleStream object. Read-only.",
     0,
-    MD_MODULESTREAM_VERSION_LATEST,
+    /* TODO: reset to MD_MODULESTREAM_VERSION_LATEST */
+    MD_MODULESTREAM_VERSION_TWO,
     0,
     G_PARAM_READABLE | G_PARAM_STATIC_STRINGS);
```
Same

mmathesius replied:
Reverted.

sgallagher commented on modulemd/tests/ModulemdTests/modulestream.py:
```diff
@@ -643,6 +647,14 @@ def test_upgrade(self):
 """,
         )
 
+    def test_upgrade_v2_to_v3(self):
+        # TODO: implement test
+        pass
+
+    def test_upgrade_v1_to_v3(self):
+        # TODO: implement test
+        pass
+
```
If the upgrade isn't implemented yet, you should probably leave the test out of the PR.

mmathesius replied:
These were placeholders to make sure I didn't forget about them. I'll include them with my upgrade work instead. I also dropped the corresponding C versions of the placeholders from this PR.

sgallagher reviewed (approved):
LGTM

mmathesius commented:
I just added one small improvement for context validation. Assuming you're good with it, can we go ahead and squash these commits and merge into the devel branch?

sgallagher commented:
> I just added one small improvement for context validation. Assuming you're good with it, can we go ahead and squash these commits and merge into the devel branch?

Go ahead.

mmathesius commented:
> > I just added one small improvement for context validation. Assuming you're good with it, can we go ahead and squash these commits and merge into the devel branch?
> 
> Go ahead.

Thanks, @sgallagher. Squashed and rebased with the OpenMandriva workaround. I'll merge when I confirm CI passes.
